File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/ArchiveTableConnection.java
package com.emogoth.android.phone.mimi.db;

import com.emogoth.android.phone.mimi.db.model.Archive;
import com.mimireader.chanlib.models.ChanArchive;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.Single;

public class ArchiveTableConnection {
    public static Single<List<Archive>> fetchArchives(String board) {
        return DatabaseUtils.fetchTable(Archive.class, Archive.TABLE_NAME, null, Archive.BOARD + "=?", board);
    }

    public static Observable<Boolean> putChanArchives(List<ChanArchive> archives) {
        return DatabaseUtils.insert(convertArchivesToDbModel(archives)).toObservable();
    }

    private static List<Archive> convertArchivesToDbModel(List<ChanArchive> archives) {
        List<Archive> models = new ArrayList<>(archives.size());

        for (ChanArchive archive : archives) {
            for (String board : archive.getBoards()) {

                Archive model = new Archive();
                model.board = board;

                model.uid = archive.getUid() == null ? -1 : archive.getUid();
                model.name = archive.getName();
                model.domain = archive.getDomain();
                model.https = archive.getHttps();
                model.software = archive.getSoftware();
                model.reports = archive.getReports() == null ? false : archive.getReports();

                models.add(model);
            }
        }

        return models;
    }

    public static Observable<Boolean> clear() {
        DatabaseUtils.WhereArg[] args = new DatabaseUtils.WhereArg[0];
        return DatabaseUtils.remove(new Archive(), true, args).toObservable();
    }
}

File: mimi-app/src/main/java/com/emogoth/android/phone/mimi/db/ArchiveTableConnection.kt
package com.emogoth.android.phone.mimi.db

import com.emogoth.android.phone.mimi.db.MimiDatabase.Companion.getInstance
import com.emogoth.android.phone.mimi.db.models.Archive
import com.mimireader.chanlib.models.ChanArchive
import io.reactivex.Single
import java.util.*

object ArchiveTableConnection {
    fun fetchArchives(board: String): Single<List<Archive>> {
        return getInstance()?.archives()?.getAllForBoard(board)?.firstOrError() ?: Single.just(emptyList())
    }

    @JvmStatic
    fun putChanArchives(archives: List<ChanArchive>): Single<Boolean> {
        return Single.defer {
            getInstance()?.archives()?.upsert(convertArchivesToDbModel(archives)) ?: return@defer Single.just(false)
            Single.just(true)
        }
    }

    private fun convertArchivesToDbModel(archives: List<ChanArchive>): List<Archive> {
        val models: MutableList<Archive> = ArrayList(archives.size)
        for (archive in archives) {
            for (board in archive.boards) {
                val model = Archive(null,
                        if (archive.uid == null) -1 else archive.uid,
                        archive.name,
                        archive.domain,
                        archive.https,
                        archive.software,
                        board, if (archive.reports == null) false else archive.reports)
                models.add(model)
            }
        }
        return models
    }

    @JvmStatic
    fun clear(): Single<Boolean> {
        return Single.defer {
            getInstance()?.archives()?.clear() ?: return@defer Single.just(false)
            Single.just(true)
        }
    }
}